pkg/operator/controller/canary/namespace.go

```diff
@@ -0,0 +1,27 @@
+package canary
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+)
+
+// ensureCanaryNamespace ensures that the ingress-canary namespace exists
+func (r *reconciler) ensureCanaryNamespace() error {
+	ns := manifests.CanaryNamespace()
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get ingress canary namespace %q: %v", ns.Name, err)
+		}
+		if err := r.client.Create(context.TODO(), ns); err != nil {
+			return fmt.Errorf("failed to create ingress canary namespace %q: %v", ns.Name, err)
```

Comment:
I think the func is easier to read and streamlined when:
```
	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
		if errors.IsNotFound(err) {
			return r.client.Create(context.TODO(), ns)
		}
                return fmt.Errorf("failed to get ingress canary namespace %q: %v", ns.Name, err)
```

The same approach can be applied to all ensure funcs.

Reply:
Interesting. My main concern with this approach is that the responsibility to log successful resource creation is pushed back to the caller. Thoughts?

Reply:
Would this be acceptable:
```
if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
    if errors.IsNotFound(err) {
        // The namespace doesn't exist, so create it.
        if err := r.client.Create(context.TODO(), ns); err != nil {
            return fmt.Errorf("failed to create ingress canary namespace %q: %v", ns.Name, err)
    }
    log.Info("created ingress canary namespace", "name", ns.Name)
    return fmt.Errorf("failed to get ingress canary namespace %q: %v", ns.Name, err)
}
```

Reply:
In your above example, `log.Info` is called to log creation success even when creation has failed. Maybe you mean something like this?
```go
if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
    if errors.IsNotFound(err) {
        // The namespace doesn't exist, so create it.
        if err := r.client.Create(context.TODO(), ns); err != nil {
            log.Info("created ingress canary namespace", "name", ns.Name)
            return fmt.Errorf("failed to create ingress canary namespace %q: %v", ns.Name, err)
    } else {
            return fmt.Errorf("failed to get ingress canary namespace %q: %v", ns.Name, err)
    }
}
```